core/src/test/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateActionTests.java

```diff
@@ -0,0 +1,855 @@
+
+        node.action.handleIncomingClusterStateRequest(new BytesTransportRequest(state1Bytes, Version.CURRENT), channel);
+        assertThat(channel.response.get(), equalTo((TransportResponse) TransportResponse.Empty.INSTANCE));
+        assertThat(channel.error.get(), nullValue());
+        channel.clear();
+
+        // another incoming state is OK. Should just override pending state
+        node.action.handleIncomingClusterStateRequest(new BytesTransportRequest(state2Bytes, Version.CURRENT), channel);
+        assertThat(channel.response.get(), equalTo((TransportResponse) TransportResponse.Empty.INSTANCE));
+        assertThat(channel.error.get(), nullValue());
+        channel.clear();
+
+        // committing previous state should fail
+        try {
+            node.action.handleCommitRequest(new PublishClusterStateAction.CommitClusterStateRequest(state1.stateUUID()), channel);
+            // sadly, there are ways to percolate errors
+            assertThat(channel.response.get(), nullValue());
+            assertThat(channel.error.get(), notNullValue());
+            if (channel.error.get() instanceof IllegalStateException == false) {
+                throw channel.error.get();
+            }
+        } catch (IllegalStateException OK) {
+
+        }
+        channel.clear();
+
+        // committing second state should succeed
+        node.action.handleCommitRequest(new PublishClusterStateAction.CommitClusterStateRequest(state2.stateUUID()), channel);
+        assertThat(channel.response.get(), equalTo((TransportResponse) TransportResponse.Empty.INSTANCE));
+        assertThat(channel.error.get(), nullValue());
+        channel.clear();
+
+        // now check it was really committed
+        assertSameState(node.clusterState, state2);
+    }
+
+    /**
+     * Tests that cluster is committed or times out. It should never be the case that we fail
+     * an update due to a commit timeout, but it ends up being committed anyway
+     */
+    public void testTimeoutOrCommit() throws Exception {
+        Settings settings = Settings.builder()
+                .put(DiscoverySettings.COMMIT_TIMEOUT, "1ms").build(); // short but so we will sometime commit sometime timeout
+
+        MockNode master = createMockNode("master", settings);
+        MockNode node = createMockNode("node", settings);
+        ClusterState state = ClusterState.builder(master.clusterState)
+                .nodes(DiscoveryNodes.builder(master.clusterState.nodes()).put(node.discoveryNode).masterNodeId(master.discoveryNode.id())).build();
+
+        for (int i = 0; i < 10; i++) {
+            state = ClusterState.builder(state).incrementVersion().build();
+            logger.debug("--> publishing version [{}], UUID [{}]", state.version(), state.stateUUID());
+            boolean success;
+            try {
+                publishState(master.action, state, master.clusterState, 2).await(1, TimeUnit.HOURS);
```

Comment:
Same out of order process as http://build-us-00.elastic.co/job/es_feature_two_phase_pub/5/testReport/junit/org.elasticsearch.discovery.zen.publish/PublishClusterStateActionTests/testTimeoutOrCommit/ . Should be fixed now.